dbt/adapters/databricks/credentials.py

```diff
@@ -241,172 +240,150 @@ def extract_cluster_id(cls, http_path: str) -> Optional[str]:
+            host=self.host,
+            client_id=self.client_id,
+            client_secret=self.client_secret,
+            auth_type="oauth-m2m"
+    )
+
+    def authenticate_with_external_browser(self):
+        return Config(
+            host=self.host,
+            client_id=self.client_id,
+            client_secret=self.client_secret,
+            auth_type="external-browser"
+    )
+
+    def authenticate_with_azure_client_secret(self):
+        return Config(
+            host=self.host,
+            azure_client_id=self.client_id,
+            azure_client_secret=self.client_secret,
+            auth_type="azure-client-secret"
+    )
+  
+    def __post_init__(self) -> None:
+        self._lock = threading.Lock()
+        with self._lock:
+            if hasattr(self, '_config') and self._config is not None:
+                # _config already exists, so skip initialization
+                return
+                
             if self.token:
-                provider = token_auth(self.token)
-                self._credentials_provider = provider.as_dict()
-                return provider
-
-            if self.client_id and self.client_secret:
-                provider = m2m_auth(
-                    host=host,
-                    client_id=self.client_id or "",
-                    client_secret=self.client_secret or "",
+                self._config = Config(
+                    host=self.host,
+                    token=self.token,
                 )
-                self._credentials_provider = provider.as_dict()
-                return provider
-
-            client_id = self.client_id or CLIENT_ID
-            redirect_url = self.oauth_redirect_url or REDIRECT_URL
-            scopes = self.oauth_scopes or SCOPES
-
-            oauth_client = OAuthClient(
-                host=host,
-                client_id=client_id,
-                client_secret="",
-                redirect_url=redirect_url,
-                scopes=scopes,
-            )
-            # optional branch. Try and keep going if it does not work
-            try:
-                # try to get cached credentials
-                credsdict = self.get_sharded_password("dbt-databricks", host)
-
-                if credsdict:
-                    provider = SessionCredentials.from_dict(oauth_client, json.loads(credsdict))
-                    # if refresh token is expired, this will throw
+            else:
+                auth_methods = {
+                    "oauth-m2m": self.authenticate_with_oauth_m2m,
+                    "azure-client-secret": self.authenticate_with_azure_client_secret,
+                    "external-browser": self.authenticate_with_external_browser
+                }
+
+                auth_type = (
+                    "external-browser" if not self.client_secret
+                    # if the client_secret starts with "dose" then it's likely using oauth-m2m
+                    else "oauth-m2m" if self.client_secret.startswith("dose")
```

Comment:
can it wrapped in a method like `is_databricks_m2m`? And this checks still a little bit tricky, do we really want to infer it from secret format? Can it be explicitly specified by the config to decide whether azure or databricks m2m should be used?

Reply:
+1 to what Jacky said.  In design review we decided to not depend on secret format, but instead to a.) reference if a config flag is set to tell us to use AAD, and b.) if we fail with Databricks OAuth but succeed with AAD, emit a warning telling them to set the config flag.